### PR TITLE
Fe 2 - Digital Cash - Creation objects Referred to the json schema 

### PR DESCRIPTION
**Pull request by zoemarin**
Creation of the objects and the tests

### CODE REVIEW

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+  public void setVersion(int version) {
+    this.version = version;
+  }
+
+  /**
+   * Function that give the Public Key of the Inputs
+   * @return List<PublicKey> senders public keys
+   */
+  public List<PublicKey> get_senders_transaction() {
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java:79]: Camel Case should be used for function names in Java

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    }
+
+    return senders;
+  }
+
+  /**
+   * Function that give the Public Key Hash of the Outputs
+   * @return List<String> outputs public keys hash
+   */
+  public List<String> get_receivers_hash_transaction() {
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java:96]: Same here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+    return receiver_hash;
+  }
+
+  /**
+   * Function that give the Public Key of the Outputs
+   * @param map_hash_key Map<String,PublicKey> dictionary public key by public key hash
+   * @return List<PublicKey> outputs public keys
+   */
+  public List<PublicKey> get_receivers_transaction(Map<String, PublicKey> map_hash_key) {
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java:112]: Same here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    return receivers;
+  }
+
+  /**
+   * Check if a public key is in the recipient
+   *
+   * @param publicKey PublicKey of someone
+   * @return true if public key in receiver, false otherwise
+   */
+  public boolean is_receiver(PublicKey publicKey) {
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java:131]: Same here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+  }
+
+  /**
+   * Function that given a key pair change the sig of an input considering all the outputs
+   *
+   * @param keyPair of one input sender
+   * @return sig other all the outputs and inputs with the public key
+   * @throws GeneralSecurityException
+   */
+  public String compute_sig_outputs_inputs(KeyPair keyPair) throws GeneralSecurityException {
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java:142]: Same here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    return keyPair.sign(new Base64URLData(String.join("", sig))).getEncoded();
+  }
+
+  /**
+   * Function that given a Public Key gives the miniLaoCoin received
+   *
+   * @param receiver Public Key of a potential receiver
+   * @return int amount of Lao Coin
+   */
+  public int get_miniLao_per_receiver(PublicKey receiver) {
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java:176]: Same here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    // TODO : Useful check ?
+    // if (!is_receiver(receiver)) {
+    // throw new IllegalArgumentException("The public Key is not contained in the receiver public
+    // key");
+    // }
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Transaction_object.java:181]: I think this should be included it is a good sanity check, but maybe have a more explicit comment

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Input_objectTest.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+import com.github.dedis.popstellar.model.objects.security.KeyPair;
+import com.github.dedis.popstellar.model.objects.security.PublicKey;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.security.GeneralSecurityException;
+
+public class Input_objectTest {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Input_objectTest.java:14]: Also should be Camel Case

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Input_objectTest.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  public void setup() throws GeneralSecurityException {
+    TYPE = "P2PKH";
+    PUBKEY = SENDER.getEncoded();
+    SIG = SENDER_KEY.sign(SENDER).getEncoded();
+    SCRIPTTXIN = new Script_input_object(TYPE, PUBKEY, SIG);
+    INPUT = new Input_object(Tx_OUT_HASH, TX_OUT_INDEX, SCRIPTTXIN);
+  }
+
+  @Test
+  public void get_tx_out_indexTest() {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Input_objectTest.java:37]: Same here

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Input_objectTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    assertEquals(TX_OUT_INDEX, INPUT.get_tx_out_index());
+  }
+
+  @Test
+  public void get_tx_out_hashTest() {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Input_objectTest.java:42]: Same here

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Input_objectTest.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    assertEquals(Tx_OUT_HASH, INPUT.get_tx_out_hash());
+  }
+
+  @Test
+  public void get_scriptTest() {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Input_objectTest.java:47]: Same here

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Output_objectTest.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateKeyPair;
+import static org.junit.Assert.assertEquals;
+
+import com.github.dedis.popstellar.model.objects.security.KeyPair;
+import com.github.dedis.popstellar.model.objects.security.PublicKey;
+
+import org.junit.Test;
+
+public class Output_objectTest {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Output_objectTest.java:11]: Same here

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Output_objectTest.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+  public static final String TYPE = "P2PKH";
+  public static String PUBKEYHASH = SENDER.computeHash();
+  private static Script_output_object SCRIPTTXOUT = new Script_output_object(TYPE, PUBKEYHASH);
+
+  private static final int VALUE = 32;
+  private static final Output_object OUTPUT = new Output_object(VALUE, SCRIPTTXOUT);
+
+  @Test
+  public void get_valueTest() {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Output_objectTest.java:23]: Same here

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Output_objectTest.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    assertEquals(VALUE, OUTPUT.get_value());
+  }
+
+  @Test
+  public void get_scriptTest() {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Output_objectTest.java:28]: Same here

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Script_input_objectTest.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    assertEquals(TYPE, SCRIPTTXIN.get_type());
+  }
+
+  @Test
+  public void getpubkeytest() {
+    assertEquals(PUBKEY, SCRIPTTXIN.get_pubkey());
+  }
+
+  @Test
+  public void getsigtest() {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Script_input_objectTest.java:43]: This is not even snake case, to be changed into Camel Case

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Script_output_objectTest.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateKeyPair;
+import static org.junit.Assert.assertEquals;
+
+import com.github.dedis.popstellar.model.objects.security.KeyPair;
+import com.github.dedis.popstellar.model.objects.security.PublicKey;
+
+import org.junit.Test;
+
+public class Script_output_objectTest {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Script_output_objectTest.java:11]: To be changed to Camel case

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Script_output_objectTest.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+  private static final KeyPair SENDER_KEY = generateKeyPair();
+  private static final PublicKey SENDER = SENDER_KEY.getPublicKey();
+
+  public static final String TYPE = "P2PKH";
+  public static String PUBKEYHASH = SENDER.computeHash();
+  private static Script_output_object SCRIPTTXOUT = new Script_output_object(TYPE, PUBKEYHASH);
+
+  @Test
+  public void get_typeTest() {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Script_output_objectTest.java:20]: Same here

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Transaction_objectTest.java [hunk 18]
```diff
@@ -0,0 +1,40 @@
+package com.github.dedis.popstellar.model.objects;
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class Transaction_objectTest {
```
Comment (Tesa27) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/objects/Transaction_objectTest.java:7]: Same here

**Review by Tesa27 — CHANGES_REQUESTED** [review]
Good job, the naming convention is not respected in almost all the functions, I think I covered them all so it is a good sanity check is after commiting the changes if y comments are outdated all the function names should be good.

**Review by Tesa27 — DISMISSED** [review]
LGTM

**Review by GabrielFleischer — APPROVED** [review]
LGTM

**Pull request merged**